pennylane_lightning/src/gates/cpu_kernels/avx_common/ApplyCRY.hpp

```diff
@@ -0,0 +1,323 @@
+// Copyright 2022 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+/**
+ * @file
+ * Defines CRY gate
+ */
+#pragma once
+#include "AVXUtil.hpp"
+#include "BitUtil.hpp"
+#include "Blender.hpp"
+#include "Permutation.hpp"
+
+#include "ConstantUtil.hpp"
+#include "Util.hpp"
+
+#include <immintrin.h>
+
+#include <complex>
+#include <utility>
+
+namespace Pennylane::Gates::AVXCommon {
+template <typename PrecisionT, size_t packed_size> struct ApplyCRY {
+    using Precision = PrecisionT;
+    using PrecisionAVXConcept = AVXConceptType<PrecisionT, packed_size>;
+
+    constexpr static auto packed_size_ = packed_size;
+    constexpr static bool symmetric = false;
+
+    /**
+     * We implement CRY gate by dividing the matrix into diagonal and
+     * off-diagonal parts. The matrix is written as: [1   0   0            0 ]
+     * [0   1   0            0           ]
+     * [0   0   cos(phi/2)   -sin(phi/2)]
+     * [0   0   sin(phi/2)   cos(phi/2)  ]
+     *
+     * Applying the matrix to a vector v, we thus
+     * (1) compute [v[0], v[1], cos(phi/2) v[2], cos(phi/2) v[3]]
+     * (2) compute [0, 0, -sin(phi/2) v[3], sin(phi/2) v[2])]
+     * and sum them.
+     *
+     * Functions related to (1) contains "Diag" in the name whereas those
+     * related to (2) contains "OffDiang".
+     * */
+
+    /**
+     * @brief Permutation for (2). Flip the target bit if control bit is 1.
+     */
+    template <size_t control, size_t target>
+    static constexpr auto applyInternalInternalPermutation() {
+        std::array<uint8_t, packed_size> perm{};
+
+        for (size_t k = 0; k < packed_size / 2; k++) {
+            if ((k >> control) & 1U) { // if control bit is 1
+                perm[2 * k + 0] = 2 * (k ^ (1U << target)) + 0;
+                perm[2 * k + 1] = 2 * (k ^ (1U << target)) + 1;
+            } else {
+                perm[2 * k + 0] = 2 * k + 0;
+                perm[2 * k + 1] = 2 * k + 1;
+            }
+        }
+        return Permutation::compilePermutation<PrecisionT>(perm);
+    }
+
+    /**
+     * @brief Factor for (2).
+     * [0, 0, 0, 0, -sin(phi/2), -sin(phi/2), sin(phi/2), sin(phi/2)]
+     */
+    template <size_t control, size_t target, class ParamT>
+    static auto applyInternalInternalOffDiagFactor(ParamT angle) {
+        std::array<PrecisionT, packed_size> arr{};
+        // positions are after permutations
+        for (size_t k = 0; k < packed_size / 2; k++) {
```

Comment:
Could it better to use bitwise shift for the division operation here?

Reply:
Integer divided two is translated to bitwise shift almost every time. https://godbolt.org/z/Kfe47rqfn

Reply:
Thanks @chaeyeunpark . Also checked that little difference between` /2 `and `*0.5` for floating number .